src/js/components/DataTable/DataTable.js

```diff
@@ -0,0 +1,138 @@
+import React, { Component } from 'react';
+import { compose } from 'recompose';
+
+import { Table } from '../Table';
+
+import { withTheme } from '../hocs';
+
+import Header from './Header';
+import Footer from './Footer';
+import Body from './Body';
+import GroupedBody from './GroupedBody';
+import buildState from './buildState';
+import doc from './doc';
+
+class DataTable extends Component {
+  static defaultProps = {
+    columns: [],
+    data: [],
+  }
+
+  state = {}
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    return buildState(nextProps, prevState);
+  }
+
+  onFiltering = (property) => {
+    this.setState({ filtering: property });
+  }
+
+  onFilter = (property, value) => {
+    const nextFilters = { ...this.state.filters };
+    nextFilters[property] = value;
+    this.setState({ filters: nextFilters });
+  }
+
+  onSort = property => () => {
+    const { sort } = this.state;
+    const ascending = (sort && property === sort.property) ? !sort.ascending : true;
+    this.setState({ sort: { property, ascending } });
+  }
+
+  onToggleGroup = groupValue => () => {
+    const groupState = { ...this.state.groupState };
+    groupState[groupValue] = {
+      ...groupState[groupValue],
+      expanded: !(groupState[groupValue].expanded),
+    };
+    this.setState({ groupState });
+  }
+
+  onToggleGroups = () => {
+    const expanded = Object.keys(this.state.groupState)
+      .filter(k => !this.state.groupState[k].expanded).length === 0;
+    const groupState = {};
+    Object.keys(this.state.groupState).forEach((k) => {
+      groupState[k] = { ...this.state.groupState[k], expanded: !expanded };
+    });
+    this.setState({ groupState });
+  }
+
+  onResize = property => (width) => {
+    const widths = { ...(this.state.widths || {}) };
+    widths[property] = width;
+    this.setState({ widths });
+  }
+
+  render() {
+    const {
+      columns, groupBy, resizeable, size, sortable, theme,
+     } = this.props;
+    const {
+      data, filtering, filters, footerValues, groups, groupState, primaryProperty,
+      showFooter, sort, widths,
+    } = this.state;
+
+    if (size && resizeable) {
+      console.warn('DataTable cannot combine "size" and "resizeble".');
+    }
+
+    return (
+      <Table>
+        <Header
+          columns={columns}
+          filtering={filtering}
+          filters={filters}
+          groups={groups}
+          groupState={groupState}
```

Comment:
it feels weird to have a prop named `groupState` since react has the concept of `state`. I think I've mentioned this in a previous PR.

Reply:
What other term would you use to track the state of the groups?